### PR TITLE
debug zip command

### LINKED ISSUE #1179: cli: add "debug zip" command
PRs: 
- https://github.com/cockroachdb/cockroach/pull/14033
- [cli: ensure that `debug zip` captures the cluster settings](https://github.com/cockroachdb/cockroach/pull/15207)

We haven't documented any `debug` commands to date, but this one needs to be documented, since it's what customers will use to collect all the info we need for troubleshooting.

### LINKED ISSUE #1372: Explain `cockroach debug zip` in docs
This new command is going to become very important for us and we will advise users to use it whenever possible. It is important that they can  understand what it does before they run it.

### PR DESCRIPTION
**Pull request by sploiselle**
Closes #1372
Closes #1179

@knz Can I get you to review for technical accuracy? If this isn't close to being right, let me know and I'd love to grab some time to discuss in greater detail. Thanks!

---
This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/docs/1469)

### CODE REVIEW

File: debug-and-error-logs.md [hunk 1]
```diff
@@ -6,16 +6,16 @@ toc: false
...
 If you need to [troubleshoot](troubleshooting-overview.html) issues with your cluster, you can check a node's logs, which include details about certain node-level and range-level events, such as errors. For example, if CockroachDB crashes, it normally logs a stack trace to what caused the problem.
 
-Each node's logs detail only the internal activity of that node without visibility into the behavior of other nodes in the cluster. When troubleshooting, this means that you must identify the node where the problem occurred or check the logs from all of the nodes in your cluster.
-
 <div id="toc"></div>
 
 ## Details
 
-Each [`cockroach` command](cockroach-commands.html) a node processes produces a stream of messages about its activities. Each message's body describes the activity, and its envelope contains metadata such as the message's severity level.
+When nodes process [`cockroach` command]s(cockroach-commands.html), they produce a stream of messages about the command's activities. Each message's body describes the activity, and its envelope contains metadata such as the message's severity level.
```
Comment (knz) [inline debug-and-error-logs.md]: I think the `s` should be inside the brackets.

File: debug-zip.md [hunk 2]
```diff
@@ -0,0 +1,88 @@
+---
+title: Collect Debug & Error Logs from Your Cluster
+summary: Learn the commands for configuring, starting, and managing a CockroachDB cluster.
+toc: false
+---
+
+The `debug zip` [command](cockroach-commands.html) connects to your cluster and gathers [log files](debug-and-error-logs.html) from each active node into a single file (inactive nodes are not included).
```
Comment (knz) [inline debug-zip.md]: `debug zip` collects much more information than just the log files.
It also includes:
- schema change events
- node liveness
- gossip data
- stack traces
- range lists
- list of databases and tables

**Comment by bdarnell** [comment]
Review status: 0 of 5 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[debug-zip.md, line 19 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/1469#-Kl5Xpi8MJZo1qcLjf68:-Kl5Xpi9DNSqreXkdltb:bfglild) ([raw file](https://github.com/cockroachdb/docs/blob/f4e69daea6e57b4a0ee868ead799c2bc9867babf/debug-zip.md#L19)):*
> ```Markdown
> Because `debug zip` relies on CockroachDB's distributed architecture, this means that nodes not currently connected to the cluster cannot respond to the request, so their log files *are not* included.
> 
> If a majority of the nodes in your cluster are live, the logs also include the data from your cluster.
> ```

The logs may include fragments of data whether a majority are up or not (and this sounds like the *entire* data is included when there is a majority, which is not true). 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/1469)*

**Comment by sploiselle** [comment]
Review status: 0 of 5 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[debug-and-error-logs.md, line 13 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/1469#-Kl0Wdfz60HUivpXRBVJ-r1-13:-KlPUfwkGZDO2Jh-dBSi:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/f4e69daea6e57b4a0ee868ead799c2bc9867babf/debug-and-error-logs.md#L13)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I think the `s` should be inside the brackets.
</blockquote></details>

Done.

---

*[debug-zip.md, line 7 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/1469#-Kl0Wdfz60HUivpXRBVK-r1-7:-KlPUgVNqfJDHIGjHuZu:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/f4e69daea6e57b4a0ee868ead799c2bc9867babf/debug-zip.md#L7)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

`debug zip` collects much more information than just the log files.
It also includes:
- schema change events
- node liveness
- gossip data
- stack traces
- range lists
- list of databases and tables
</blockquote></details>

Done.

---

*[debug-zip.md, line 19 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/1469#-Kl5Xpi8MJZo1qcLjf68:-KlPUgr02YsHuxZH86fa:b9h4mzv) ([raw file](https://github.com/cockroachdb/docs/blob/f4e69daea6e57b4a0ee868ead799c2bc9867babf/debug-zip.md#L19)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

The logs may include fragments of data whether a majority are up or not (and this sounds like the *entire* data is included when there is a majority, which is not true). 
</blockquote></details>

Can I get more information about which/when data is included?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/1469)*

**Comment by knz** [comment]
Reviewed 3 of 5 files at r1, 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/1469)*

**Comment by bdarnell** [comment]
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[debug-zip.md, line 19 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/1469#-Kl5Xpi8MJZo1qcLjf68:-KlPixfft5AqXH6YkWjd:bndmiqm) ([raw file](https://github.com/cockroachdb/docs/blob/f4e69daea6e57b4a0ee868ead799c2bc9867babf/debug-zip.md#L19)):*
<details><summary><i>Previously, sploiselle (Sean Loiselle) wrote…</i></summary><blockquote>

Can I get more information about which/when data is included?
</blockquote></details>

Not really, it's kind of unpredictable. We don't specifically try to include any data here, but sometimes bits of data appear in the logs (especially for values that happen to be chosen as range split keys). For example, in https://github.com/cockroachdb/cockroach/issues/16008#issuecomment-302545177, the first "initiating a split" line contains the primary key of one row. In this case it's just a random number in a test database, but depending on the schema it could include something like an email address or other private information. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/1469)*

**Comment by sploiselle** [comment]
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[debug-zip.md, line 19 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/1469#-Kl5Xpi8MJZo1qcLjf68:-KlPxRfMPEMu4FiWntkp:bijrdp5) ([raw file](https://github.com/cockroachdb/docs/blob/f4e69daea6e57b4a0ee868ead799c2bc9867babf/debug-zip.md#L19)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Not really, it's kind of unpredictable. We don't specifically try to include any data here, but sometimes bits of data appear in the logs (especially for values that happen to be chosen as range split keys). For example, in https://github.com/cockroachdb/cockroach/issues/16008#issuecomment-302545177, the first "initiating a split" line contains the primary key of one row. In this case it's just a random number in a test database, but depending on the schema it could include something like an email address or other private information. 
</blockquote></details>

Got it. I'll place an oblique mention of that in the warning.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/1469)*

**Comment by sploiselle** [comment]
@jseldess Ready for copy edit.

File: debug-and-error-logs.md [hunk 3]
```diff
@@ -6,16 +6,16 @@ toc: false
...
 If you need to [troubleshoot](troubleshooting-overview.html) issues with your cluster, you can check a node's logs, which include details about certain node-level and range-level events, such as errors. For example, if CockroachDB crashes, it normally logs a stack trace to what caused the problem.
 
-Each node's logs detail only the internal activity of that node without visibility into the behavior of other nodes in the cluster. When troubleshooting, this means that you must identify the node where the problem occurred or check the logs from all of the nodes in your cluster.
-
 <div id="toc"></div>
 
 ## Details
 
-Each [`cockroach` command](cockroach-commands.html) a node processes produces a stream of messages about its activities. Each message's body describes the activity, and its envelope contains metadata such as the message's severity level.
+When nodes process [`cockroach` commands](cockroach-commands.html), they produce a stream of messages about the command's activities. Each message's body describes the activity, and its envelope contains metadata such as the message's severity level.
```
Comment (jseldess) [inline debug-and-error-logs.md]: ni, for count consistency/clarity: 

>When a node process a `cockroach` command, it produces a stream of messages...

File: debug-and-error-logs.md [hunk 4]
```diff
@@ -6,16 +6,16 @@ toc: false
...
 
 As commands generate messages, CockroachDB uses the [command](#commands)'s [logging flags](#flags) and the message's [severity level](#severity-levels) to determine the appropriate [location](#output-locations) for it.
```
Comment (jseldess) [inline debug-and-error-logs.md]: nit, continued: 
> As a command generates messages...

File: debug-zip.md [hunk 5]
```diff
@@ -0,0 +1,98 @@
+---
+title: Collect Debug Information from Your Cluster
+summary: Learn the commands for configuring, starting, and managing a CockroachDB cluster.
```
Comment (jseldess) [inline debug-zip.md]: Let's make the summary specific to this page.

File: debug-zip.md [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+
+<div id="toc"></div>
+
+## Details
+
+### Use Cases
+
+There are two scenarios in which `debug zip` is useful:
+
+- To collect all of your node's logs, which you can then parse to locate issues. It's important to note, though, that `debug zip` can only access logs from active nodes. See more information [on this page](#collecting-log-files).
```
Comment (jseldess) [inline debug-zip.md]: `nodes' logs`

File: debug-zip.md [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+
+### Collecting Log Files
+
+When you issue the `debug zip` command, the node that receives the request connects to each other node in the cluster. Once it's connected, the node requests the content of all log files stored on the node, the location of which is determined by the `--log-dir` value when you [started the node](start-a-node.html).
+
+Because `debug zip` relies on CockroachDB's distributed architecture, this means that nodes not currently connected to the cluster cannot respond to the request, so their log files *are not* included.
+
+After receiving the log files from all of the active nodes, the requesting node aggregates the files and writes them to an archive file you specify.
+
+You can locate logs in the unarchived file's`debug/nodes/[node dir]/logs` directories.
```
Comment (jseldess) [inline debug-zip.md]: Need a space after `file's`.

File: debug-zip.md [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+~~~ shell
+# Generate a debug zip:
+$ cockroach debug zip [ZIP file destination] [flags]
+~~~
+
+It's important to understand that the `[flags]` here are used to connect to CockroachDB nodes. This means the values you use in those flags must connect to an active node. If no nodes are live, you must [start at least one node](start-a-node.html).
+
+## Flags
+
+The `debug zip` subcommand subcommand supports the following [general-use](#general) and [logging](#logging) flags.
```
Comment (jseldess) [inline debug-zip.md]: Remove extra `subcommand`.

**Comment by sploiselle** [comment]
Thanks for the quick review, @jseldess. Ended up putting the doc under **Troubleshoot** per your suggestion.

---

Review status: 3 of 5 files reviewed at latest revision, 7 unresolved discussions.

---

*[debug-and-error-logs.md, line 13 at r3](https://reviewable.io:443/reviews/cockroachdb/docs/1469#-Kl0Wdfz60HUivpXRBVJ-r3-13:-KlZVKx_eFpfwD3we2Ur:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/638482bd83ee9f7f2704c3fe39847eb37fb47810/debug-and-error-logs.md#L13)):*
<details><summary><i>Previously, jseldess (Jesse Seldess) wrote…</i></summary><blockquote>

nit, for count consistency/clarity: 

>When a node processes a `cockroach` command, it produces a stream of messages...
</blockquote></details>

Done.

---

*[debug-and-error-logs.md, line 15 at r3](https://reviewable.io:443/reviews/cockroachdb/docs/1469#-Kl0Wdfz60HUivpXRBVJ-r3-15:-KlZVPXERLz-6vDF-MVU:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/638482bd83ee9f7f2704c3fe39847eb37fb47810/debug-and-error-logs.md#L15)):*
<details><summary><i>Previously, jseldess (Jesse Seldess) wrote…</i></summary><blockquote>

nit, continued: 
> As a command generates messages...
</blockquote></details>

Done.

---

*[debug-zip.md, line 3 at r3](https://reviewable.io:443/reviews/cockroachdb/docs/1469#-Kl0Wdfz60HUivpXRBVK-r3-3:-KlZVUv3ZkFGYKpebrAo:b-c1susb) ([raw file](https://github.com/cockroachdb/docs/blob/638482bd83ee9f7f2704c3fe39847eb37fb47810/debug-zip.md#L3)):*
<details><summary><i>Previously, jseldess (Jesse Seldess) wrote…</i></summary><blockquote>

Let's make the summary specific to this page.
</blockquote></details>

Derp. Good catch. Thanks.

---

*[debug-zip.md, line 27 at r3](https://reviewable.io:443/reviews/cockroachdb/docs/1469#-Kl0Wdfz60HUivpXRBVK-r3-27:-KlZVe2mvIhJ0gI0DkV3:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/638482bd83ee9f7f2704c3fe39847eb37fb47810/debug-zip.md#L27)):*
<details><summary><i>Previously, jseldess (Jesse Seldess) wrote…</i></summary><blockquote>

`nodes' logs`
</blockquote></details>

Done.

---

*[debug-zip.md, line 41 at r3](https://reviewable.io:443/reviews/cockroachdb/docs/1469#-Kl0Wdfz60HUivpXRBVK-r3-41:-KlZVmzUQOsw8RhbZjgv:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/638482bd83ee9f7f2704c3fe39847eb37fb47810/debug-zip.md#L41)):*
<details><summary><i>Previously, jseldess (Jesse Seldess) wrote…</i></summary><blockquote>

Need a space after `file's`.
</blockquote></details>

Done.

---

*[debug-zip.md, line 60 at r3](https://reviewable.io:443/reviews/cockroachdb/docs/1469#-Kl0Wdfz60HUivpXRBVK-r3-60:-KlZVpDhbKtAf34w7kl5:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/638482bd83ee9f7f2704c3fe39847eb37fb47810/debug-zip.md#L60)):*
<details><summary><i>Previously, jseldess (Jesse Seldess) wrote…</i></summary><blockquote>

Remove extra `subcommand`.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/1469#-:-KlZY6jtCKXiFBmRHpWJ:bl519r7)*

**Comment by jseldess** [comment]
*[debug-and-error-logs.md, line 15 at r3](https://reviewable.io:443/reviews/cockroachdb/docs/1469#-Kl0Wdfz60HUivpXRBVJ-r3-15:-Kl_9dpb3620gFI0tNnx:b-c4ckq0) ([raw file](https://github.com/cockroachdb/docs/blob/638482bd83ee9f7f2704c3fe39847eb37fb47810/debug-and-error-logs.md#L15)):*
<details><summary><i>Previously, sploiselle (Sean Loiselle) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Don't see the change here either...

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/1469)*

**Comment by jseldess** [comment]
*[debug-and-error-logs.md, line 13 at r3](https://reviewable.io:443/reviews/cockroachdb/docs/1469#-Kl0Wdfz60HUivpXRBVJ-r3-13:-Kl_9c5l-HaEBMgiM4M-:b-s4jspo) ([raw file](https://github.com/cockroachdb/docs/blob/638482bd83ee9f7f2704c3fe39847eb37fb47810/debug-and-error-logs.md#L13)):*
<details><summary><i>Previously, sploiselle (Sean Loiselle) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Hmm, I don't see the change...

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/1469)*

**Comment by jseldess** [comment]
*[debug-and-error-logs.md, line 13 at r3](https://reviewable.io:443/reviews/cockroachdb/docs/1469#-Kl0Wdfz60HUivpXRBVJ-r3-13:-Kl_F1J7DL_Cev3_rjP2:bb6ywmx) ([raw file](https://github.com/cockroachdb/docs/blob/638482bd83ee9f7f2704c3fe39847eb37fb47810/debug-and-error-logs.md#L13)):*
<details><summary><i>Previously, jseldess (Jesse Seldess) wrote…</i></summary><blockquote>

Hmm, I don't see the change...
</blockquote></details>

LGTM

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/1469)*

**Comment by jseldess** [comment]
:lgtm:

---

Review status: 3 of 6 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[debug-and-error-logs.md, line 15 at r3](https://reviewable.io:443/reviews/cockroachdb/docs/1469#-Kl0Wdfz60HUivpXRBVJ-r3-15:-Kl_F5UpMny1MIxMzfP3:bb6ywmx) ([raw file](https://github.com/cockroachdb/docs/blob/638482bd83ee9f7f2704c3fe39847eb37fb47810/debug-and-error-logs.md#L15)):*
<details><summary><i>Previously, jseldess (Jesse Seldess) wrote…</i></summary><blockquote>

Don't see the change here either...
</blockquote></details>

LGTM

---

*[debug-zip.md, line 3 at r3](https://reviewable.io:443/reviews/cockroachdb/docs/1469#-Kl0Wdfz60HUivpXRBVK-r3-3:-Kl_FBSmV4hwikjv-xyz:bb6ywmx) ([raw file](https://github.com/cockroachdb/docs/blob/638482bd83ee9f7f2704c3fe39847eb37fb47810/debug-zip.md#L3)):*
<details><summary><i>Previously, sploiselle (Sean Loiselle) wrote…</i></summary><blockquote>

Derp. Good catch. Thanks.
</blockquote></details>

LGTM

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/1469#-:-Kl_FRsPmfoCsM27mtZS:bnfp4nl)*

**Pull request merged**